### PR TITLE
Optimise finding type of plists

### PR DESCRIPTION
**Pull request by ChrisJefferson**
Try to speed up finding the type of plists.

This particularly speeds up lists of integers, by around 5 times. In general it speeds up by 20-30%

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
Example improvement:

```
gap> l := List([1..100000], i->i+1);;
gap> while Length(l)>0 do Remove(l,Length(l)); od; time;
```

Goes from 50 seconds to 3 seconds (lists of integers are the best case).

**Comment by fingolfin** [comment]
This sounds pretty cool!

Of course the code example you give is a bit insane to start with (expanding a range into a full list) -- but of course one can easily construct (or likely, find) equivalent situations where it just can't be easily avoided, at least without adding lots of extra code which deals separately with ranges.

**Comment by frankluebeck** [comment]
Of course, any speed-up  is fine. But this doesn't solve the general problem  that the determination of the type of a list takes far too much time. In particular, more interesting cases like
```
gap> l := List([1..100000], i-> [i]);;
gap> while Length(l)>0 do Remove(l,Length(l)); od; time;
gap> l := List([1..100000], i-> [i, [i]]);;
gap> while Length(l)>0 do Remove(l,Length(l)); od; time;
```
do not profit much from this change.

**Comment by ChrisJefferson** [comment]
I agree, I did this while trying to understand the slowdown, but I think cases like this are only fixable by not scanning the list at all (which will require a bigger change)

File: src/plist.c [hunk 1]
```diff
@@ -277,11 +280,14 @@ Int KTNumPlist (
     else {
 	if (!testing) MARK_LIST(list, TESTING);
 
-	if (IS_PLIST(elm))
-	  family = FAMILY_TYPE( TypePlistWithKTNum(elm, &ktnumFirst));
+	if (IS_PLIST(elm)) {
+      typeObj = TypePlistWithKTNum(elm, &ktnumFirst);
+	  family = FAMILY_TYPE( typeObj );
+    }
```
Comment (fingolfin) [inline src/plist.c]: The indention here is inconsistent (tabs vs. spaces)

File: src/plist.c [hunk 2]
```diff
@@ -277,11 +280,14 @@ Int KTNumPlist (
...
 	else
 	  {
-	    family  = FAMILY_TYPE( TYPE_OBJ(elm) );
+        typeObj =  TYPE_OBJ(elm);
+	    family  = FAMILY_TYPE( typeObj );
 	    ktnumFirst = 0;
```
Comment (fingolfin) [inline src/plist.c]: tabs vs spaces

File: src/plist.c [hunk 3]
```diff
@@ -313,15 +328,33 @@ Int KTNumPlist (
             isHom   = 0;
             areMut  = (areMut || IS_MUTABLE_PLIST(elm));
             isTable = 0;
-	    isRect = 0;
+	        isRect = 0;
```
Comment (fingolfin) [inline src/plist.c]: tabs vs. spaces?

**Comment by fingolfin** [comment]
Other than minor indentation issues, I see nothing wrong with this case. Sure, it only helps with a limited things of examples, but as long as it does that, and makes nothing else worse (and I don't think it does), this is an improvement.

So, I'd be happy to merge this -- but perhaps @ChrisJefferson can fix the indentation first (it was inconsistent in that function before this PR already, but some of the changes make it worse)

**Comment by ChrisJefferson** [comment]
Have fixed the indentation, yes this file is a mess, this makes the file locally consistent at least...

**Pull request merged**